flang/lib/Lower/OpenMP.cpp

```diff
@@ -2988,6 +2989,24 @@ genTargetOp(Fortran::lower::AbstractConverter &converter,
            .getIsTargetDevice())
     cp.processNowait(nowaitAttr);
 
+  if (outerCombined) {
+    for (const Fortran::parser::OmpClause &clause : clauseList.v) {
+      if (const auto &reductionClause =
+              std::get_if<Fortran::parser::OmpClause::Reduction>(&clause.u)) {
+        const auto &objectList{
+            std::get<Fortran::parser::OmpObjectList>(reductionClause->v.t)};
+
+        for (const Fortran::parser::OmpObject &ompObject : objectList.v) {
+          if (const auto *name{
+                  Fortran::parser::Unwrap<Fortran::parser::Name>(ompObject)}) {
+            if (const Fortran::semantics::Symbol * symbol{name->symbol})
+              redSymbols.push_back(symbol);
+          }
+        }
+      }
+    }
+  }
```

Comment:
I think it might be better to move the logic here to gather the list of reduction symbols to a new `ClauseProcessor::processTargetReduction()` method so it follows a bit more closely the process for all other clauses.

Reply:
I tried to add processTargetReduction but it seemed like quite a bit of code duplication. Instead I now use existing processReduction to get the list of variables. That seems cleaner and there is no duplication. But we are not comparing mlir::Value instead of Symbol*. Also I have to do this on the baseOp to make the comparison work and I am not sure what this code does.

if (auto t = baseOp.getDefiningOp<hlfir::DeclareOp>())
          declOp = t.getBase();

The tests work ok. I am just wondering if it is safe to compare mlir::Value.

Reply:
My concern with calling `ClauseProcessor::processReduction()` here is that it does other things and not just gather a list of symbols, so it's not quite the same as what you wrote here. That is, it makes changes to the MLIR representation by adding the reduction declarations and new MLIR ops.

If the most of the code here is already present in `ClauseProcessor::processReduction()`, maybe a good middle ground would be to extract that part into a helper function to reuse there and in the new `ClauseProcessor::processTargetReduction()`.